integration/tests/cook/test_multi_user.py

```diff
@@ -252,6 +252,75 @@ def test_rate_limit_while_creating_job(self):
+        with user:
+            jobs = []
+            try:
+                jobspec = {"command":"sleep 240", 'cpus': 0.03, 'mem': 32}
+                # Submit 9 jobs. They should all launch.
+                jobs1a, resp1a = util.submit_jobs(self.cook_url, jobspec, 9)
+                jobs.extend(jobs1a)
+                self.assertEqual(201,resp1a.status_code, msg=resp1a.content)
+                # In testing, scheduler loop is every second.
+                time.sleep(5.0)
+
+                # Submit 9 more jobs. They may all launch, arriving in queue at once.
+                jobs1b, resp1b = util.submit_jobs(self.cook_url, jobspec, 9)
+                jobs.extend(jobs1b)
+                self.assertEqual(201,resp1b.status_code, msg=resp1b.content)
+                time.sleep(40.0)
+
+                # Submit 10 more jobs. They should be launch rate limited.
+                jobs2, resp2 = util.submit_jobs(self.cook_url, jobspec, 10)
+                jobs.extend(jobs2)
+                self.assertEqual(201,resp2.status_code, msg=resp2.content)
+                time.sleep(45.0)
+
+                one_hour_in_millis = 60 * 60 * 1000
+                start = util.current_milli_time() - one_hour_in_millis
+                end = util.current_milli_time()
+
+                resp_waiting = util.jobs(self.cook_url, user=user.name, start=start, end=end, state='waiting')
+                self.assertEqual(200,resp_waiting.status_code, msg=resp_waiting.content)
+                resp_running = util.jobs(self.cook_url, user=user.name, start=start, end=end, state='running')
+                self.assertEqual(200,resp_running.status_code, msg=resp_running.content)
+
+                number_jobs_waiting = len(resp_waiting.json())
+                number_jobs_running= len(resp_running.json())
+
+                # We expect up to 18 jobs to be running, because second batch of 9 could 'slip in'
+                # in the entirety because we're not in debt.
+                self.assertGreaterEqual(number_jobs_running,9)
+                self.assertLessEqual(number_jobs_running,18)
+                self.assertGreaterEqual(number_jobs_waiting,10)
+                self.assertLessEqual(number_jobs_waiting,19)
+
+                # Look at unscheduled.
+                unscheduled, _ = util.unscheduled_jobs(self.cook_url, *jobs)
+
+                # Is one of the reasons this job isn't launched because we're launch rate limited?
+                is_job_launch_rate_limited = [any([reasons.JOB_LAUNCH_RATE_LIMIT == reason['reason'] for reason in ii['reasons']]) for ii in unscheduled]
+                # How many are in that state?
+                num_launch_rate_limited = len([ii for ii in is_job_launch_rate_limited if ii])
+                self.assertGreaterEqual(num_launch_rate_limited,10)
+                self.assertLessEqual(num_launch_rate_limited,19)
+            finally:
+                util.kill_jobs(self.cook_url, jobs)
+
+
+
```

Comment:
As discussed offline, can we run the integration test with a specific "rate-limit" user so that it won't affect other tests that may need to use the same user?

Reply:
That problem doesn't occur here. The only config that doesn't cause this test to be skipped is a specific config configured for the travis environment. With that config (rate=5/minute, bucket=10), most of our integration tests would fail, whether we made this change or not.

Its a relevant concern, but I'd rather solve it by putting the submission rate limit integration test into its own travis environment, isolating its configuration from the other integration tests.

Reply:
It would be nice to add a comment at the top of the test explaining that repeated runs of the test with the same user will fail if sufficient time has not passed since the last run